common/json_writer/src/lib.rs

```diff
@@ -0,0 +1,130 @@
+use std::fmt::Write;
+
+#[derive(Debug, Clone)]
+pub struct JsonWriter {
+    buffer: String,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum JsonWriterError {
+    #[error("Failed to write the f64 value")]
+    Writef64ValueError(#[from] std::fmt::Error),
+}
+
+impl JsonWriter {
+    pub fn new() -> Self {
+        Self {
+            buffer: String::new(),
+        }
+    }
+    pub fn with_capacity(capa: usize) -> Self {
+        Self {
+            buffer: String::with_capacity(capa),
+        }
+    }
+
+    pub fn write_key_noescape(&mut self, key: &str) {
+        self.write_str_noescape(key);
+        self.buffer.push(':');
+    }
+
+    pub fn write_str_noescape(&mut self, s: &str) {
+        self.buffer.push('"');
+        self.buffer.push_str(s);
+        self.buffer.push('"');
+    }
+
+    pub fn write_f64(&mut self, value: f64) -> Result<(), JsonWriterError> {
+        Ok(self.buffer.write_fmt(format_args!("{}", value))?)
+    }
```

Comment:
As said above about the `JsonWriterError`, if would prefer to hide this error.

```
 Ok(self.buffer.write_fmt(format_args!("{}", value)).expect("write_fmt cannot fail on a string buffer"))
```

I do think this is one of the values of crate like this one: to hide details. If the phantom error is exposed, then all callers will have to deal with it.

Reply:
I think this is a reasonable place to call `expect()` but I don't understand why we are writing our own JSON serialiser. What's wrong with: https://docs.rs/json/0.12.4/json/#serialize-with-jsonstringifyvalue

Reply:
Remember, if we do add `expect()` then it needs to be documented as an exception to our security policies.

Reply:
Or https://stackoverflow.com/questions/59047280/how-to-build-json-arrays-or-objects-dynamically-with-serde-json

Reply:
@ChrisGreenaway TBH, I didn't know about `json::stringify`. But, there
are good reasons NOT to use `json::stringify`:

* `json::stringify` is rather inefficient for writing strings, as it will allocate it the string. And this is really really performance critical.
* we are using crate `serde_json` and not `json`.
* Unfortunately `serde_json` does not expose it's json serializer/writer as public. If it were public, I'd like to use that.

Reply:
And `serde_json::Map` is allocating too. All our Visitors are trying to be non-allocating and if we now drop in "allocation" just for serialization, all our gains are lost :/

Reply:
I had a look through the `json::stringify` code and it looks quite similar to this PR - except that it correctly handles escaping and arrays and so on. In terms of allocation, perhaps you can point out where the extra overhead is?

The place I can see this called from is `thin_edge_json::json::parse_str` which does a `json::parse` - so aren't we using `json` rather than `serde_json`?

On the performance critical nature of this code, there are several points:
1. It's the end-to-end performance that matters (i.e. from topic to topic) rather than the performance of individual pieces of code in that pipeline. What end-to-end measurements do we have of this code vs more battle-tested crates?
2. Faster is not necessarily better. It comes at the cost of more code and more bugs.
3. We can expend engineering effort to make things faster, but is it worth it? Are we already fast enough? How do we decide that code needs to be faster?